### ISSUE TITLE
drush sqlc doesn't work reliably with the Drush Launcher or globally-managed Drush

### ISSUE THREAD

**Issue by anavarre** [drush-ops/drush#3309]
## Context

I have a binary and 2 aliases under `/usr/local/bin`

```
$ (cd /usr/local/bin/ && ls -l *drush*)
-rwxr-xr-x 1 root root  274420 Jan 13 22:30 drush
lrwxrwxrwx 1 root staff     46 Jan 15 10:14 drush9 -> /var/www/html/drucker/vendor/drush/drush/drush
lrwxrwxrwx 1 root staff     48 Jan 15 10:20 global-drush -> /home/drucker/.composer/vendor/drush/drush/drush
```

* `drush` is the Drush Launcher (0.5.1)
* `drush9` is my vendored Drush (RC2)
* `global-drush` is a global Composer-managed Drush (RC2)

I have a site alias which looks like this:

```
$ drush sa
'@drucker.local':
  root: /var/www/html/drucker
  uri: 'http://drucker.local'
```

And in the context of the Drush Launcher, I have an env var that defaults to the `global-drush` alias.

```
export DRUSH_LAUNCHER_FALLBACK=`which global-drush``
```

Note: all tests are performed with Docker containers, where the DB has its own container. See https://github.com/anavarre/drucker for details.

## Problem

When using the `sqlc` command with the `--root` parameter, it works immediately, but only with the `drush` (Drush Launcher) and `drush9` (vendored Drush) commands. When using the `global-drush` command (global Composer-managed Drush) it just hangs forever until I cancel the command.

```
$ drush --root=/var/www/html/drucker/ sqlc
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 154
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]> exit
Bye
$ drush9 --root=/var/www/html/drucker/ sqlc
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 158
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]> exit
Bye
drucker@drucker:/$ global-drush --root=/var/www/html/drucker/ sqlc
^C
```

When using a site alias, invoking `drush` (Drush Launcher) or `global-drush` (globally-managed Drush), it hangs forever. It only works when invoking `drush9` (site-local Drush).

```
$ drush @drucker.local sqlc
^C
drucker@drucker:/$ drush9 @drucker.local sqlc
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 170
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]> exit
Bye
drucker@drucker:/$ global-drush @drucker.local sqlc
^C
```

## Observations

* Whether an alias is used or not, the result is different, even when passing the `--root` parameter.
* There's no way to easily understand what's going on when it hangs forever as the `-vvv` or `-vd` parameter don't seem to be working with Drush 9
* Since the alias has a `root` key, there should be no difference when running `drush @drucker.local` or `drush --root=/var/www/html/drucker`
* Other commands bootstrapping Drupal and connecting to the DB such as `sqlq` work well in every scenario (alias or `--root` parameter, with either `drush`, `drush9` or `global-drush`)

**Comment by weitzman** [comment]
I'm not sure whats going on.

However, in your nice Docker setup, I would remove the global drush9 and the drush launcher and just keep the asite local drush. In the container, edit so that vendor/bin/drush is in your $PATH. Then you can just call `drush` and be done with it. Further, you dont have to pass --root because drush9 is smart enough to find root when it is located inside vendor.

**Comment by anavarre** [comment]
> I would remove the global drush9 and the drush launcher and just keep the asite local drush.

Initially this is what I had done to follow your recommendation to now go with a vendored Drush. But, to try and think about a way you could rely on one-Drush-to-rule-them-all on a server/container with N number of sites potentially using deprecated Drush versions (imagine `composer.json` files requiring Drush 8, still), the global Drush is somewhat appealing.

Will try and see if I can troubleshoot the issue and, if not, I'll probably follow your suggestion.

Thanks, Moshe!

**Comment by jakubsuchy** [comment]
I think there are plenty of use cases for global drush that this use case should not be forgotten. If you only rely on your vendor drush you might not be able to guarantee certain commands to succeed if the codebase is not under your control. 

My team has another use case - we have a site tarball that's reused over and over again and a bash script that relies on drush before it decompresses the tarball - in a no global drush scenario it would fail quickly...

**Comment by weitzman** [comment]
1. Drush project can't fix this. Its a Composer/PHP design choice. See
https://pantheon.io/blog/trouble-two-autoloaders
2. OK, so untar and then proceed as normal. Or install a global Drush and
hope for the best.

**Comment by anavarre** [comment]
Here's more data about `sqlc` hanging forever. It hangs just after:

```
 [debugnotify] Add a commandfile class: Drush\Drupal\Commands\sql\SanitizeUserTableCommands [0.14 sec, 17.99 MB]
Calling proc_open(mysql --defaults-file=/tmp/drush_Tz01aS --database=drucker --host=203.0.113.12 --port=3306);
```

Temp file seems legit:

```
$ cat /tmp/drush_Tz01aS
#This file was written by Drush's Sqlmysql.php.
[client]
user="root"
password="root"
```

The same exact command with the creds passed at the command line works immediately:

```
$ mysql --database=drucker --host=203.0.113.12 --port=3306 -uroot -proot
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 70
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]>
```

I've removed the Drush Launcher and the global Drush to limit the annoyance factor. As recommended earlier, I've also added `/var/www/html/drucker` to my `$PATH`, which allows to invoke Drush nicely indeed.

```
$ echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/drucker/.composer/vendor/bin:/home/drucker/.composer/vendor/bin:/home/drucker/.composer/vendor/bin:/var/www/html/drucker/vendor/bin/
```

I'm stumped as to why `sql:dump`, `sqlq` and other commands would work, but not `sqlc`.

**Comment by blkperl** [comment]
I am able to reproduce the sqlc issue. Strace shows it hanging on the read call

```
[pid   167] rt_sigaction(SIGINT, {0x409e20, [INT], SA_RESTORER|SA_RESTART, 0x7fecfbb084b0}, {SIG_DFL, [], 0}, 8) = 0
[pid   167] rt_sigaction(SIGQUIT, {0x409a90, [QUIT], SA_RESTORER|SA_RESTART, 0x7fecfbb084b0}, {SIG_DFL, [], 0}, 8) = 0
[pid   167] rt_sigaction(SIGHUP, {0x409e20, [HUP], SA_RESTORER|SA_RESTART, 0x7fecfbb084b0}, {SIG_DFL, [], 0}, 8) = 0
[pid   167] sendto(3, "!\0\0\0\3select @@version_comment li"..., 37, 0, NULL, 0) = 37
[pid   167] recvfrom(3, "\1\0\0\1\1'\0\0\2\3def\0\0\0\21@@version_comme"..., 16384, 0, NULL, NULL) = 99
[pid   167] read(0,
```

**Comment by anavarre** [comment]
Ok, it's not only me then. Reopening.

**Comment by segi** [comment]
I could reproduce the problem as well as you @anavarre wrote down above.

**Comment by anavarre** [comment]
@weitzman that's 3 of us able to replicate the `sqlc` command hanging forever. How would you suggest we make progress on the issue going forward?

**Comment by weitzman** [comment]
I suspect that there is a bug in php's proc_open() on some systems. From your log:
```
proc_open(mysql --defaults-file=/tmp/drush_Tz01aS --database=drucker --host=203.0.113.12 --port=3306);
```
Does that fail if you run it with `php -r`? If so you have a boiled down failure case. Try it with --user and --password instead if `--defaults-file=/tmp/drush_Tz01aS`. I suspect that the hang is unrelated to --defaults-file but I'm not sure. If my hunch is correct, its time to consult the google -
 https://www.google.com/search?q=proc_open+hangs

**Comment by anavarre** [comment]
Followed your advice.

```
$ php -a
Interactive mode enabled

php > $sql = 'mysql --user=root --password=root --database=drucker --host=203.0.113.12 --port=3306';
php > proc_open($sql, [], $pipes, NULL, NULL, ['bypass_shell' => TRUE]);
php > Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 384
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]>

And with a 'creds' file like Drush uses:

```
$ php -a
Interactive mode enabled

php > $sql = 'mysql --defaults-file=creds --database=drucker --host=203.0.113.12 --port=3306';
php > proc_open($sql, [], $pipes, NULL, NULL, ['bypass_shell' => TRUE]);
php > Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 385
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]>
```

Oh well.

**Comment by weitzman** [comment]
Are you guys all using same version of PHP? Maybe post output of `drush status`. I'm pretty much out of ideas.

**Comment by anavarre** [comment]
No we're not. They're likely using PHP 7.1.11 on Ubuntu and I'm using 7.2 on Debian 9 as shown below.

```
PHP 7.2.1-1+0~20180105151615.16+stretch~1.gbpd3910a (cli) (built: Jan  5 2018 15:16:17) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.2.1-1+0~20180105151615.16+stretch~1.gbpd3910a, Copyright (c) 1999-2017, by Zend Technologies
```

FYI I'm back to the original setup when I started the issue, when experimenting with the Launcher and a (test) global Drush to rule them all.

Looking at `drush status`, I noticed the following.

```
drucker@drucker:/$ drush status
 PHP binary    : /usr/bin/php7.2                                      
 PHP config    : /etc/php/7.2/cli/php.ini                             
 PHP OS        : Linux                                                
 Drush script  : /usr/local/bin/global-drush                          
 Drush version : 9.0.0-rc2                                            
 Drush temp    : /tmp                                                 
 Drush configs : /home/drucker/.drush/drush.yml                       
                 /home/drucker/.composer/vendor/drush/drush/drush.yml 
drucker@drucker:/$ global-drush status
 PHP binary    : /usr/bin/php7.2                                      
 PHP config    : /etc/php/7.2/cli/php.ini                             
 PHP OS        : Linux                                                
 Drush script  : /usr/local/bin/global-drush                          
 Drush version : 9.0.0-rc2                                            
 Drush temp    : /tmp                                                 
 Drush configs : /home/drucker/.drush/drush.yml                       
                 /home/drucker/.composer/vendor/drush/drush/drush.yml 
drucker@drucker:/$ drush9 status
 Drupal version   : 8.6.0-dev                                          
 Site URI         : default                                            
 DB driver        : mysql                                              
 DB hostname      : 203.0.113.12                                       
 DB port          : 3306                                               
 DB username      : root                                               
 DB name          : drucker                                            
 Database         : Connected                                          
 Drupal bootstrap : Successful                                         
 Default theme    : bartik                                             
 Admin theme      : seven                                              
 PHP binary       : /usr/bin/php7.2                                    
 PHP config       : /etc/php/7.2/cli/php.ini                           
 PHP OS           : Linux                                              
 Drush script     : /usr/local/bin/drush9                              
 Drush version    : 9.0.0                                              
 Drush temp       : /tmp                                               
 Drush configs    : /home/drucker/.drush/drush.yml                     
                    /var/www/html/drucker/vendor/drush/drush/drush.yml 
 Install profile  : standard                                           
 Drupal root      : /var/www/html/drucker/docroot                      
 Site path        : sites/default                                      
 Files, Public    : sites/default/files                                
 Files, Temp      : /tmp
```

As you can see above, MySQL discovery without an alias only seems to work in the context of the recommended setup which is the vendored Drush I've added in my $PATH. When I'm passing either `--root` or an alias, I get always get MySQL discovery just fine, which is why I can run commands such as `sqlq` just fine.

**Comment by webflo** [comment]
drush-launcher 0.6.0 contains a fix for interactive commands (https://github.com/drush-ops/drush-launcher/pull/44). Maybe this could help to resolve the issue.

**Comment by anavarre** [comment]
Thanks @webflo but unfortunately it didn't (at least for me).

* `drush` is the Drush Launcher (0.6.0)
* `drush9` is my vendored Drush (9.2.1)
* `global-drush` is a global Composer-managed Drush (9.1.0)

```
$ ls -l /usr/local/bin/*drush*
-rwxr-xr-x 1 root root    310920 Mar  9 18:01 drush
lrwxrwxrwx 1 root staff       47 Mar  9 18:11 drush9 -> /var/www//html/drucker/vendor/drush/drush/drush
lrwxrwxrwx 1 root staff       48 Feb 12 14:34 global-drush -> /home/drucker/.composer/vendor/drush/drush/drush
$ drush --root=/var/www/html/drucker/ --version | head -n1 | awk '{print $NF}'
0.6.0
$ global-drush --root=/var/www/html/drucker/ sqlc
^C
$ drush @drucker.local sqlc
^C
$ drush9 --root=/var/www/html/drucker/ sqlc
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 56
Server version: 10.1.26-MariaDB-0+deb9u1 Debian 9.1

Copyright (c) 2000, 2017, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [drucker]>
```

So, exact same issue as previously.

**Comment by weitzman** [comment]
Fixed in drush 9.6.
